File: ruby/024/024.rb
class Array
  def nth_permutation(n)
    raise ArgumentError if n > (1..length).inject(:*)

    sorted = if block_given?
               sort do |x, y|
                 yield x, y
               end
             else
               sort
             end

    sorted_with_indices = sorted.map.with_index { |element, index| [index, element] }.to_h

    result = []

    until sorted_with_indices.empty?
      m = 0

      f = if sorted_with_indices.length > 1
            (1..(sorted_with_indices.length - 1)).inject(:*)
          else
            1
          end

      m += 1 while m * f < n

      m -= 1

      result << sorted_with_indices[m]

      sorted_with_indices.delete(m)

      n -= m * f

      sorted_with_indices = sorted_with_indices.values.map.with_index { |element, index| [index, element] }.to_h
    end

    result
  end
end

digits = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

puts digits.nth_permutation(1000000).join

File: elixir/024/024.exs
defmodule Euler do
  def nth_permutation(list, n) do
    nth_permutation(list |> Enum.sort |> Enum.with_index, [], n)
  end

  defp nth_permutation([_ | _] = list, current, max) do
    f = cond do
      length(list) > 1 -> (1..(length(list) - 1)) |> Enum.reduce(1, &(&1 * &2))
      true -> 1
    end

    index = get_index(0, f, max)

    result = current ++ [elem(:lists.nth(index + 1, list), 0)]

    nth_permutation(Enum.map(List.delete_at(list, index), fn {v, _} -> v end) |> Enum.with_index, result, max - (index * f))
  end

  defp nth_permutation([], current, _) do
    current
  end

  defp get_index(m, f, max) when m * f < max do
    get_index(m + 1, f, max)
  end

  defp get_index(m, _, _) do
    m - 1
  end
end

IO.puts [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
|> Euler.nth_permutation(1000000)
|> Enum.map(&(to_string(&1)))
|> Enum.join
